open_feature/provider/in_memory_provider.py

```diff
@@ -0,0 +1,122 @@
+from dataclasses import dataclass
+import typing
+
+from open_feature.evaluation_context.evaluation_context import EvaluationContext
+from open_feature.exception.error_code import ErrorCode
+from open_feature.flag_evaluation.reason import Reason
+from open_feature.flag_evaluation.resolution_details import FlagResolutionDetails
+from open_feature.hooks.hook import Hook
+from open_feature.provider.metadata import Metadata
+from open_feature.provider.provider import AbstractProvider
+
+PASSED_IN_DEFAULT = "Passed in default"
+
+
+@dataclass
+class InMemoryMetadata(Metadata):
+    name: str = "In-Memory Provider"
+
+
+T = typing.TypeVar("T", covariant=True)
+
+
+@dataclass
+class InMemoryFlag(typing.Generic[T]):
+    flag_key: str
+    default_variant: str
+    variants: typing.Dict[str, T]
+    reason: typing.Optional[Reason] = Reason.STATIC
+    error_code: typing.Optional[ErrorCode] = None
+    error_message: typing.Optional[str] = None
```

Comment:
I don't think we should embed error code & message into the flag structure. These values should be resolved from evaluation logic 

For example, consider this go example [1]. This is the in-memory search operation and if flag is not found based on flag key, this returns a flag not found [2].


[1] - https://github.com/open-feature/go-sdk/pull/192/files#diff-ebcfeaa6fba781be983c41afae4094c9324c532721a8bbb7b6a03cebaf37d027R132-R143
[2] - https://github.com/open-feature/spec/blob/main/specification/types.md#error-code

Reply:
@federicobond my initial thought of in-memory provider was mainly on tests. However, some discussions and community opinions on OFEP [1] [2] suggests that this can live outside of tests. 

With that background, I think we should have some sort of basic error handling built-in to this provider 

[1] - https://github.com/open-feature/ofep/pull/71#discussion_r1263230706
[2] - https://github.com/open-feature/ofep/pull/71#issuecomment-1638384086